x-pack/plugins/observability_solution/observability/public/components/alerts_flyout/alert_flyout_overview/helpers/map_rules_params_with_flyout.ts

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import {
+  ALERT_EVALUATION_VALUE,
+  ALERT_EVALUATION_VALUES,
+  ALERT_RULE_PARAMETERS,
+  ALERT_RULE_TYPE_ID,
+  METRIC_INVENTORY_THRESHOLD_ALERT_TYPE_ID,
+  OBSERVABILITY_THRESHOLD_RULE_TYPE_ID,
+  LOG_THRESHOLD_ALERT_TYPE_ID,
+  ALERT_EVALUATION_THRESHOLD,
+  ApmRuleType,
+  SLO_BURN_RATE_RULE_TYPE_ID,
+} from '@kbn/rule-data-utils';
+import { EsQueryRuleParams } from '@kbn/stack-alerts-plugin/public/rule_types/es_query/types';
+import { i18n } from '@kbn/i18n';
+import { asDuration, asPercent } from '../../../../../common';
+import { createFormatter } from '../../../../../common/custom_threshold_rule/formatters';
+import { metricValueFormatter } from '../../../../../common/custom_threshold_rule/metric_value_formatter';
+import { METRIC_FORMATTERS } from '../../../../../common/custom_threshold_rule/formatters/snapshot_metric_formats';
+import { METRIC_THRESHOLD_ALERT_TYPE_ID } from '../../../../pages/alert_details/alert_details';
+import {
+  BaseMetricExpressionParams,
+  CustomMetricExpressionParams,
+} from '../../../../../common/custom_threshold_rule/types';
+import { TopAlert } from '../../../../typings/alerts';
+import { isFieldsSameType } from './is_fields_same_type';
+export interface FlyoutThresholdData {
+  observedValue: string;
+  threshold: string[];
+  comparator: string;
+  pctAboveThreshold: string;
+}
+
+const getPctAboveThreshold = (observedValue?: number, threshold?: number[]): string => {
+  if (!observedValue || !threshold) return '';
+  if (threshold.length > 1) {
+    return i18n.translate('xpack.observability.alertFlyout.overview.rangeThresholdLabel', {
+      defaultMessage: ' (Range threshold)',
+    });
+  }
+  return i18n.translate('xpack.observability.alertFlyout.overview.aboveThresholdLabel', {
+    defaultMessage: ' ({pctValue}% above the threshold)',
+    values: {
+      pctValue: (((observedValue - threshold[0]) * 100) / threshold[0]).toFixed(2),
+    },
+  });
+};
+
+export const mapRuleParamsWithFlyout = (alert: TopAlert): FlyoutThresholdData[] | undefined => {
+  const ruleParams = alert.fields[ALERT_RULE_PARAMETERS];
+  if (!ruleParams) return;
+  const ruleCriteria = ruleParams?.criteria as Array<Record<string, any>>;
+  const ruleId = alert.fields[ALERT_RULE_TYPE_ID];
+  const observedValues: number[] = alert.fields[ALERT_EVALUATION_VALUES]! || [
+    alert.fields[ALERT_EVALUATION_VALUE]!,
+  ];
+
+  switch (ruleId) {
+    case OBSERVABILITY_THRESHOLD_RULE_TYPE_ID:
```

Comment:
@fkanout @benakansara Sorry for this late comment, about having rule-specific logic in the observability plugin, could we consider moving this logic to each rule instead?

A concern here is that by having this logic separate from where the rule is defined, it can get out of sync when something in the rule changes, and having it closer to the rule definition makes it easier to update this logic in that scenario. (It also helps developers to have a better overview of rule-related codes when all of it is in one place.)

I am thinking about having a similar approach as the `alertDetailsAppSection` that each rule provides. (Something like `getFlyoutParams` with a shared interface that is provided by each rule.)

Just to be on the same page: my question is about where is the place for this logic, and how part of it (this PR, or a separate PR) can be done in different ways (not the concern of this comment).

Reply:
@maryam-saeidi, I agree this would be a good solution. Why not use this file as a baseline in the following PR!
However, here are some points I thought about that we need to keep in mind and drove me away from doing it in this PR:
- When I started this, the scope was narrow, covering CTR and Metric rules, but it became wider over time to cover all the Obs rules.
- I didn't know the required field for each rule to build the flyout. 
- Ideally, we would not need this file, `map_rules_params_with_flyout.ts`. Alerts should have the same document structure, especially when it comes to these essential fields. This may be why the current implementation is done in one place. So, the right call could be pushing toward unifying these fields across alerts.

Reply:
@fkanout Thanks for clarification.

> Ideally, we would not need this file, map_rules_params_with_flyout.ts. Alerts should have the same document structure, especially when it comes to these essential fields. This may be why the current implementation is done in one place. So, the right call could be pushing toward unifying these fields across alerts.

It would be great if we could implement this approach with a reasonable amount of effort. Based on what you learned, do you think it is possible to have this approach for all the fields or is there going to be still some fields that are dependent on the rule type for specific treatment?

Reply:
@maryam-saeidi I think it is feasible even if there are some edge cases.
Listing everything here is hard, but having this map in one file will make identifying the differences/similarities easier. 
But I can list some examples: e.g., Comparator will significantly help if it's exposed on the alert level. APM-ErrorCount and SLO are identical regarding the fields required for the flyout. Align the comparator of Logs with other rules.